ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Adversarial.hs

```diff
@@ -0,0 +1,536 @@
+        let pc = BV.countActivesInV S.notInverted vA
+        when (C.toVar pc <= 0) $ Exn.throwError BadCount
+
+    -- the youngest slot in the adversarial schedule that cannot have accelerated
+    --
+    -- (IE s past the first active adversarial slot.)
+    youngestStableA :: C.Index adv SlotE
+    youngestStableA =
+        case BV.findIthEmptyInV S.inverted vA (C.Count 0) of
+            BV.JustFound firstActiveA ->
+                -- if s=0, then the slot of their block is the youngest stable slot
+                firstActiveA C.+ s
+            BV.NothingFound           ->
+                -- the rest of the function won' force this since there are no
+                -- adverarial active slots
+                error "dead code"
+
+    checkRaces = do
+        let iterH =
+                maybe (RI.initConservative (Scg s) (Delta d) winH) id
+              $ RI.init (Kcp k) vH
+        case RI.init (Kcp (k - 1)) vA >>= RI.next vA of
+            Nothing    -> pure ()   -- there are <= k total adversarial active slots
+            Just iterA ->
+                -- TODO optimization: how to skip all the honest Race Windows that
+                -- don't reach beyond the intersection? Perhaps skip to i - s + d?
+                go iterH iterA
+
+    -- INVARIANT iterH spans k+1 active slots (possibly conservatively)
+    --
+    -- INVARIANT iterA spans k active slots (actually, not conservatively)
+    go !iterH !iterA = do
+        C.SomeWindow Proxy raceWinH <- pure $ let RI.Race x = iterH in x
+        C.SomeWindow Proxy raceWinA <- pure $ let RI.Race x = iterA in x
+
+        -- lift both windows to @base@ so that they're comparable
+        let raceWinH' = C.joinWin winH raceWinH
+            raceWinA' = C.joinWin winA raceWinA
+
+        if
+
+          -- any Race Window that ends /after/ the adversary can accelerate is unconstrained
+          | C.frWin winA youngestStableA < C.windowLast raceWinH' -> pure ()
+
+          -- advance the adversarial Race Window if its start is <= the honest Race Window's start
+          | C.windowStart raceWinA' <= C.windowStart raceWinH' ->
+            case RI.next vA iterA of
+                Just iterA' -> go iterH iterA'
+                Nothing     -> pure ()   -- there are < k remaining adversarial active slots
+
+          -- fail if the adversary won or tied the race
+          | C.toVar (C.windowLast raceWinA') <= C.toVar (C.windowLast raceWinH') + C.Count d ->
+                -- iterA contains exactly k active slots, but A) it's anchored
+                -- in an active slot and B) iterH contains that anchor. Thus
+                -- adv has k+1 in iterH.
+                Exn.throwError $ BadRace AdversaryWonRace {
+                    rvAdv = iterA
+                  ,
+                    rvHon = iterH
+                  }
+
+          -- advance the honest Race Window
+          | otherwise -> case RI.next vH iterH <|> RI.nextConservative (Scg s) (Delta d) vH iterH of
```

Comment:
Since "Races before acceleration lower bound" doesn't require checking windows that spill over the end of time, conservative race windows aren't necessary here.

Reply:
This counter-example is produced if we just remove conservative race windows from adversarial schema generation.

k=5, s=14, d=1
```
0111100101000001111
        11111111110
```

We need to propose how to change the generation to stop examples like these from being produced, and we still need to check that they aren't produced.

Reply:
I don't think we will remove the conservative race windows after more discussion.